src/content/docs/en/guides/actions.mdx

```diff
@@ -560,24 +550,155 @@ const result = Astro.getActionResult(actions.addToCart);
+  if (action?.calledFrom === "form") {
+    const actionResult = await action.handler();
+  
+    // Persist the action result using session storage
+    const sessionId = randomUUID();
+    await actionStore.setJSON(sessionId, {
+      actionName: action.name,
+      actionResult: serializeActionResult(actionResult),
+    });
+  
+    // Pass the session ID as a cookie
+    // to be retrieved after redirecting to the page
+    context.cookies.set("action-session-id", sessionId);
+  
+    // Redirect back to the previous page on error
+    if (actionResult.error) {
+      const referer = context.request.headers.get("Referer");
+      if (!referer) {
+        throw new Error(
+          "Internal: Referer unexpectedly missing from Action POST request.",
+        );
+      }
+      return context.redirect(referer);
+    }
+    // Redirect to the destination page on success
+    return context.redirect(context.originPathname);
+  }
+  
+  return next();
+});
+```
+
+## Security when using actions
+
+Actions are accessible as public endpoints based on the name of the action. For example, the action `blog.like()` will be accessible from `/_actions/blog.like`. This is useful for unit testing action results and debugging production errors. However, this means you **must** use same authorization checks that you would consider for API endpoints and on-demand rendered pages.
+
+### Authorize users from an action handler
+
+The easiest way to authorize action requests is to add an authentication check to your action handler. You may reach for [an authentication library](https://docs.astro.build/en/guides/authentication/) to handle session management and user information. Actions expose the full `APIContext` object to access properties passed from middleware using `context.locals`. When a user is not authorized, you can raise an `ActionError` with the `UNAUTHORIZED` code:
+
+```ts title="src/actions/index.ts" {6-8}
+import { defineAction, ActionError } from 'astro:actions';
 
 export const server = {
-  addToCart: defineAction({
-    handler: async () => {
-      /* ... */
-      return product;
-      return { productName: product.name };
+  getUserSettings: defineAction({
+    handler: async (_input, context) => {
+      if (!context.locals.user) {
+        throw new ActionError({ code: 'UNAUTHORIZED' });
+      }
+      return { /* data on success */ };
     }
   })
 }
 ```
-:::
+
+### Gate actions from middleware
+
+<p><Since v="5.0.0" /></p>
+
+Astro recommends authorizing user sessions from your action handler to respect permission levels and rate-limiting on a per-action basis. Still, you may want to gate requests to all actions (or a subset of actions) from middleware. 
+
+Use the `getActionContext()` function from your middleware to retrieve information about any inbound action requests. This including the action name and whether that action was called using a client-side RPC function (example: `actions.blog.like()`) or an HTML form.
```

Comment:
```suggestion
Use the `getActionContext()` function from your middleware to retrieve information about any inbound action requests. This includes the action name and whether that action was called using a client-side remote procedure call (RPC) function (e.g. `actions.blog.like()`) or an HTML form.
```
In general, we use the full name outside the parentheses and the abbreviation inside, but obviously we don't do this for EVERY thing (HTML, CSS) but we DO do it for MOST things. If you think this rises to the level of HTML/CSS, then we can consider not name checking it at all, or perhaps swapping and putting the full name in parentheses to show we have reader empathy. I'll need you to make the call here.